app/policies/result_policy.rb

```diff
@@ -52,4 +67,9 @@ def download?
       )
     )
   end
+
+  # This is a separate policy function because it reports a specific error message on failure
+  def view_with_result_token?
+    check?(:member?, record.submission.grouping) && record.view_token == view_token && !record.view_token_expired?
```

Comment:
Okay I looked into this a bit more. The `has_secure_token :view_token` sets the token on create, so new results have one set, but old results don't. So when I was testing this PR on an existing assignment, I see a bunch of `nil`s.

It's not a deal-breaker for this semester, but I think this could be a good time to try out a slightly more sophisticated migration that sets `view_token`s for all existing results (and then puts a non-null restriction on the column). We haven't done migrations like that before but I don't think it would be too bad, and that way you get to try that out and we get an extra constraint we want in the database.

On the other hand, I'm not sure if you were originally planning for cases where `view_token` would be `nil`? (E.g., only generate tokens when results are released.) Happy to talk more!